python/cucim/src/cucim/skimage/filters/_median.py

```diff
@@ -79,5 +112,64 @@ def median(image, footprint=None, out=None, mode='nearest', cval=0.0,
 
     if footprint is None:
         footprint = ndi.generate_binary_structure(image.ndim, image.ndim)
+
+    if algorithm == 'sorting':
+        can_use_histogram = False
+    elif algorithm in ['auto', 'histogram']:
+        can_use_histogram, reason = _can_use_histogram(image, footprint)
+    else:
+        raise ValueError(f"unknown algorithm: {algorithm}")
+
+    if algorithm == 'histogram' and not can_use_histogram:
+        raise ValueError(
+            "The histogram-based algorithm was requested, but it cannot "
+            f"be used for this image and footprint (reason: {reason})."
+        )
+
+    # The sorting-based implementation in CuPy is faster for small footprints.
+    # Empirically, shapes above (13, 13) and above on RTX A6000 have faster
+    # execution for the histogram-based approach.
+    use_histogram = can_use_histogram
+    if algorithm == 'auto':
+        # prefer sorting-based algorithm if footprint shape is small
+        use_histogram = use_histogram and math.prod(footprint.shape) > 150
+
+    if use_histogram:
+        try:
+            # as in SciPy, a user-provided `out` can be an array or a dtype
+            output_array_provided = False
+            out_dtype = None
+            if out is not None:
+                output_array_provided = isinstance(out, cp.ndarray)
+                if not output_array_provided:
+                    try:
+                        out_dtype = cp.dtype(out)
+                    except TypeError:
+                        raise TypeError(
+                            "out must be either a cupy.array or a valid input "
+                            "to cupy.dtype"
+                        )
+
+            # TODO: Can't currently pass an output array into _median_hist as a
+            #       new array currently needs to be created during padding.
+            temp = _median_hist(image, footprint, mode=mode, cval=cval,
+                                **algorithm_kwargs)
+            if output_array_provided:
+                out[:] = temp
+            else:
+                if out_dtype is not None:
+                    temp = temp.astype(out_dtype, copy=False)
+                out = temp
+            return out
+        except KernelResourceError as e:
+            # Fall back to sorting-based implementation if we encounter a
+            # resource limit (e.g. insufficient shared memory per block).
+            print(e)
```

Comment:
Should this `print` still be here?

Reply:
Good catch. I have now changed it to a warning instead. If the user wants to avoid the warning in the future they could then set the `algorithm='sorting'` kwarg to bypass trying the histogram-based algorithm.